common/src/stack/command/stack/commands/set/host/power/imp_ssh.py

```diff
@@ -0,0 +1,41 @@
+# @copyright@
+# Copyright (c) 2006 - 2019 Teradata
+# All rights reserved. Stacki(r) v5.x stacki.com
+# https://github.com/Teradata/stacki/blob/master/LICENSE.txt
+# @copyright@
+
+import stack.commands
+import shlex
+from stack.util import _exec
+from stack.exception import CommandError
+
+class Implementation(stack.commands.Implementation):
+	def run(self, args):
+		host = args[0]
+		cmd = args[1]
+
+		if cmd == 'on':
+			raise CommandError(self, f'Cannot use ssh to start host {host}')
+
+		if cmd == 'off':
+			cmd_output = _exec(shlex.split(f'ssh {host} "shutdown -h now"'))
+			out = cmd_output.stdout
+			err = cmd_output.stderr
+			if cmd_output.returncode != 0:
+				raise CommandError(self, err)
+
+		if cmd == 'reset':
+			cmd_output = _exec(shlex.split(f'ssh {host} "reboot"'))
+			out = cmd_output.stdout
+			err = cmd_output.stderr
+			if cmd_output.returncode != 0:
+				raise CommandError(self, err)
+
+		if cmd == 'status':
+			cmd_output = _exec(shlex.split(f'ssh {host} "hostname"'))
+			out = cmd_output.stdout
+			err = cmd_output.stderr
+			if cmd_output.returncode != 0:
+				raise CommandError(self, f'Chassis Power is unreachable via ssh:\n{err}')
+			else:
+				return f'Chassis Power is on'
```

Comment:
This command is inconsistent in that sometimes it returns nothing but in other cases it returns a string. This should be made consistent in every code branch.

Reply:
Ah so besides when raising a command error, a string should always be returned?

Reply:
It depends on what this is supposed to do. I don't have a strong opinion one way or the other, it just needs to be consistent.

Reply:
The idea was messages that are supposed to be run without debug enabled would be returned and command errors would be included in debug messages in the main command __init__ file.

Reply:
I actually disagree with @ryancausey here.  Typically unix commands shouldn't output any text unless there's a problem or it's specifically requested (which is what is happening in the above logic).

I would add, however, that you may be leaving out useful diagnostic output by only returning STDERR in the command error.  Rather than returning both (as output might become unordered) you can pipe them together.

```
import subprocess

_exec('hostname -garbage'.split(), stderr=subprocess.STDOUT, stdout=subprocess.PIPE)
```

STDERR will be piped to STDOUT, and STDOUT will have all output, in the same order it would have appeared in the terminal.  A cleaner way to specify that in `stack.util._exec` that didn't require importing subprocess would be welcome.

Also, `_exec` should already be part of the command class, which means you should be able to get to it without import via `self._exec` or `self.owner._exec`

Reply:
@bsanders I'm not talking about the command printing text or not, I'm saying that _this function_ should have a consistent return API. It should either always return some value (string, return code, whatever) or always return nothing. In most of the `if/elif` cases there is no return value, whereas in the last one there is a return of a string. For what it is worth, linters will also complain about the inconsistent return value behavior.

To be clear I'm specifically talking about `return` statements. Exceptions should still be thrown in error cases as necessary.

Reply:
Ah, I see.  So since the message is needed in one case, return an empty string for the others.

Reply:
I'm returning a three element named tuple always now that consists of `(console_output, debug_output, imp_succeed)` where the last element is a boolean if the implementation succeeded or not.